File: Online Judges/LeetCode/July LeetCoding Challenge/Week-2/3Sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        Ans = []
        nums.sort()

        for i in range(len(nums)):
            if i == 0 or (i > 0 and nums[i-1] != nums[i]):
                low = i + 1
                high = len(nums) - 1
                extra = - nums[i]
                while low < high:
                    if nums[low] + nums[high] == extra:
                        Ans.append([nums[i], nums[low], nums[high]])

                        while low < len(nums)-1 and nums[low] == nums[low+1]:
                            low += 1
                        while high > 0 and nums[high] == nums[high-1]:
                            high -= 1
                    if (nums[low] + nums[high]) < extra:
                        low += 1
                    else:
                        high -= 1
        return Ans

File: Online Judges/LeetCode/July LeetCoding Challenge/Week-2/3Sum.rs
impl Solution {
    pub fn three_sum(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ans = Vec::new();
        
        let n = nums.len() as usize;
        
        let mut nums = nums;
        nums.sort();
        
        for i in 0..n {
            if i==0 || (i>0 && nums[i-1] != nums[i]) {
                
                let mut low = i + 1;
                let mut high = nums.len() - 1;
                let extra =  - nums[i];
                
                while low < high {
                    if (nums[low] + nums[high]) == extra{
                        
                        let v = vec![nums[i], nums[low], nums[high]];
                        ans.push(v);
                    
                        while (low < high) && (nums[low] == nums[low+1]){
                            low += 1;
                        }
                        while ((high > low) && (nums[high] == nums[high - 1])){
                            high -= 1;
                        }
                    }
                    if extra > (nums[low] + nums[high]){
                        low +=1;
                    }
                    else{
                        high -= 1;
                    }
                }
            }
        }
        return ans;
    }
}